### PR TITLE
Rotate certificate upon valid principals change.

### PR DESCRIPTION
**Pull request by russjones**
**Description**

Ignore 0.0.0.0 when checking if certificate needs to be rotated.

### CODE REVIEW

File: lib/service/connect.go [hunk 1]
```diff
@@ -600,41 +600,57 @@ type rotationStatus struct {
...
 
-	additionalPrincipals, dnsNames, err := process.getAdditionalPrincipals(id.Role)
-	if err != nil {
-		return nil, trace.Wrap(err)
+	// If the list of principals contains 0.0.0.0 (meaning advertise_ip has not
+	// been set), remove it from the list of principals. The 0.0.0.0 values tells the
+	// auth server to "guess" the nodes IP. Since this IP is not known (possibly
+	// even to the node itself if it's behind some kind of NAT), checking if it's
+	// in the list of principals would not make sense.
+	principalToCheck := additionalPrincipals
```
Comment (klizhentas) [inline lib/service/connect.go]: can we just always non-conditionally exclude 0.0.0.0 from the slice, would not it make the logic simpler?

File: lib/service/connect.go [hunk 2]
```diff
@@ -600,6 +600,37 @@ type rotationStatus struct {
...
+	// list of principals in the certificate. Since it never exists in the list
+	// of principals (auth server will always remove it before issuing a
+	// certificate) regeneration is always requested.
+	principalsToCheck := utils.RemoveFromSlice(additionalPrincipals, defaults.AnyAddress)
+
+	// If advertise_ip, public_addr, or listen_addr in file configuration were
+	// updated, the list of principals (SSH) or DNS names (TLS) on the
+	// certificate need to be updated.
+	if len(additionalPrincipals) != 0 && !conn.ServerIdentity.HasPrincipals(principalsToCheck) {
+		principalsChanged = true
```
Comment (klizhentas) [inline lib/service/connect.go]: if all you care about is that the certificates have changed, you can return here, and avoid the extra variable?

File: lib/service/connect.go [hunk 3]
```diff
@@ -600,6 +600,37 @@ type rotationStatus struct {
...
+		log.Debugf("Rotation in progress, adding %v to SSH principals %v.",
+			additionalPrincipals, conn.ServerIdentity.Cert.ValidPrincipals)
+	}
+	if len(dnsNames) != 0 && !conn.ServerIdentity.HasDNSNames(dnsNames) {
+		dnsNamesChanged = true
```
Comment (klizhentas) [inline lib/service/connect.go]: Ditto here?

File: lib/service/connect.go (hunk 2 above)
Comment (russjones) [reply lib/service/connect.go]: It's for logging purposes, so we can debug what changed (was it SSH or TLS or both).

Comment (klizhentas) [reply lib/service/connect.go]: gotcha

**Review by klizhentas — APPROVED** [review]

**Pull request merged**